File: RB101/exercises/medium_2/tri_angles.rb
=begin
testing the angles of a triangle to return type of triangle

Rules:
1. angles add up to 180
2. no angles can be 0
3. Types
  - Right - one angle equals 90
  - Acute - all angles less than 90
  - Obtuse - one angle greater than 90

=end

def triangle(a1, a2, a3)
  angles = [a1, a2, a3]
  return :invalid if angles.any?(0) || angles.sum != 180

  if angles.any?(90)
    :right
  elsif angles.all? { |a| a < 90 }
    :acute
  elsif angles.any? { |a| a > 90 }
    :obtuse
  else
    :invalid
  end
end

def triangle(a1, a2, a3)
  angles = [a1, a2, a3]
  return :invalid if angles.any?(0) || angles.sum != 180

  test = angles.select { |a| a >= 90 }[0]
  return :acute if test.nil?
  case test > 90
  when true then :obtuse
  else
    :right
  end
end

def triangle(a1, a2, a3)
  angles = [a1, a2, a3]
  return :invalid if angles.any?(0) || angles.sum != 180

  test = angles.select { |a| a >= 90 }[0]

  case 
  when test.nil?
    :acute
  when test > 90
    :obtuse
  when test == 90
    :right
  end
end

p triangle(60, 70, 50) == :acute
p triangle(30, 90, 60) == :right
p triangle(120, 50, 10) == :obtuse
p triangle(0, 90, 90) == :invalid
p triangle(50, 50, 50) == :invalid

File: LS215/exercises/medium2/tri_angles.js
function triangle(a1, a2, a3) {
  let angles = [a1, a2, a3];

  if ((angles.reduce((sum, a) => a + sum) !== 180) || angles.some(a => a === 0)) {
    return 'invalid';
  }
  
  if (angles.every(a => a < 90)) {
    return 'acute';
  } else {
    let oneOver90 = 0;
    let one90 = 0;
    angles.forEach(a => {
      if (a > 90) {
        oneOver90 += 1;
      } else if (a === 90) {
        one90 += 1;
      }
    });

    if (oneOver90 === 1 && one90 === 0) {
      return 'obtuse';
    } else if (oneOver90 === 0 && one90 === 1) {
      return 'right';
    } else {
      return 'invlaid';
    }
  }
}

console.log(triangle(60, 70, 50));       // "acute"
console.log(triangle(30, 90, 60));       // "right"
console.log(triangle(120, 50, 10));      // "obtuse"
console.log(triangle(0, 90, 90));        // "invalid"
console.log(triangle(50, 50, 50));       // "invalid"